mitodl/open-discussions
# Updated user api to create social auth if provider_username is present

#### What are the relevant tickets?
Fixes #967 

#### What's this PR do?
This PR takes a few steps to handling `UserSocialAuth` where the `uid` is the MicroMasters username:

- Updates the user/profile API to accept an optional `uid` (optional for backwards compat). If it's passed, the API creates/updates the username.
- Updates the `BaseJwtAuth` backend to more accurately process the JWT. It does so by passing the JWT payload as `response`, so that the core PSA implementation will populate `uid` based on the looking up `ID_KEY` in the response. 
- Updates the JWT token handling to look up the username if a `provider` is provided. MM will later be updated to pass this *only for user-facing JWTs*. Backend JWTs will not have this so we also preserve the old behavior both for that, and to ensure that we can time delivery of the MM changes separately.

#### How should this be manually tested?

- You could make a test call to the API, but since it's just another optional field this isn't strictly necessary
- You should still be able to authenticate via JWT to the discussions API as a user
- You'll need to test with the MM branch `nl/social_auth_tokens`. If you create a `UserSocialAuth` record for your user with `provider='micromasters', uid='$MM_USERNAME'` you should be able to update MM `OPEN_DISCUSSIONS_REDIRECT_URL` to point to `/complete/micromasters` and see that properly give you a session for that user. **NOTE:** the MM username needs to be `User.username`, this is *not* the same as `SETTINGS.user.username`.
